mautic/mautic
# Allow plugin create own deduplicate logic

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Common `mautic:contacts:deduplicate` already deduplicate contacts by unique fields from configurations. But sometimes we need custom check for duplications (for example based on firstname/lastname/phone). This PR added dispatcher to checkForDuplicateContacts method allow load duplicate contacts by plugin. 
This allow also create own command (for example `mautic:contacts:deduplicate:custom`) along with core one. 


#### Steps to test this PR:
1. Install and setup MauticCustomDeduplicateBundle https://github.com/mtcextendee/mautic-custom-deduplicate-bundle
2. Choose in plugins settings:
firstname, lastname, phone
and not empty email
3. Create two contacts without email but with same firstname, lastname, and phone.
4. Run `php app/console mautic:contacts:deduplicate:custom`
5. See If contacts were merged
6. Then create two contacts again
7. Go to Mautic > Contacts
8. Run Custom deduplicate from menu
![image](https://user-images.githubusercontent.com/462477/57359025-f7dc4b80-7176-11e9-94ed-eb0dee9e0c24.png)
9. Wait and then see If you have at least two notification in notification area (started, finished)
![image](https://user-images.githubusercontent.com/462477/57359081-12aec000-7177-11e9-92ff-e77991f0e875.png)
10. Check If contacts were merged
